client/asset/eth/eth.go

```diff
@@ -3779,53 +3829,67 @@ func (w *assetWallet) sumPendingTxs() (out, in uint64) {
 		}
 		if pt.lastCheck == tip {
 			// Expect nothing has changed since our last check.
-			addPendingSend(pt)
-			continue
-		}
-		confs, err := w.node.transactionConfirmations(w.ctx, txHash)
-		if err != nil && !errors.Is(err, asset.CoinNotFoundError) {
-			w.log.Errorf("Error getting confirmations for pending tx %s: %v", txHash, err)
-			// DRAFT NOTE: Should I stop tracking this tx? I'd hate to end up in
-			// a state where we're spamming logs every block for a tx that will
-			// never happen, though presumably that would result in
-			// CoinNotFoundError.
-			// delete(w.pendingTxs, txHash)
-			continue
-		}
-		if confs > 0 {
-			delete(w.pendingTxs, txHash)
+			addPendingTx(pt)
 			continue
 		}
-		pt.lastCheck = tip // Avoid multiple checks on the same block.
-		addPendingSend(pt)
+		wg.Add(1)
+		go updatePendingTx(txHash, pt)
+	}
+	wg.Wait()
+	for txHash := range deletions {
+		delete(w.pendingTxs, txHash)
 	}
 	return
 }
 
 func (w *assetWallet) balanceWithTxPool() (*Balance, error) {
 	isETH := w.assetID == BipID
-	var confirmed *big.Int
-	var err error
-	if isETH {
-		confirmed, err = w.node.addressBalance(w.ctx, w.addr)
-	} else {
-		confirmed, err = w.tokenBalance()
+
+	getBal := func() (*big.Int, error) {
+		if isETH {
+			return w.node.addressBalance(w.ctx, w.addr)
+		} else {
+			return w.tokenBalance()
+		}
 	}
+
+	confirmed, err := getBal()
+
 	if err != nil {
 		return nil, fmt.Errorf("balance error: %v", err)
 	}
 
-	mempoolNode, is := w.node.(mempoolFetcher)
+	txPoolNode, is := w.node.(txPoolFetcher)
 	if !is {
 		out, in := w.sumPendingTxs()
+
+		// To avoid the possibility that a pending spending tx is mined after we
+		// pull the balance but before we check pending txs, which would result
+		// in an incorrectly high balance, we'll play safe here and check the
+		// balance again.
+		checkBal, err := getBal()
+		if err != nil {
+			return nil, fmt.Errorf("balance consistency check error: %v", err)
+		}
+		outEVM := w.evmify(out)
+		// If our balance has gone down in the interim, we'll use the lower
+		// balance, but ensure that we're not setting up an underflow for
+		// available balance.
+		if checkBal.Cmp(confirmed) < 0 {
+			confirmed = checkBal
```

Comment:
Worth it verdict: possibly not, esp. if we still just deal with a before/after balance change by possibly erroring.

I figure'd we'd go do it all again if `checkBal.Cmp(confirmed) != 0`, not use the new balance that is likely to hit the error just below.

```go
for {
	confirmed, err := getBal()
	out, in = w.sumPendingTxs()
	checkBal, err := getBal()
	if checkBal.Cmp(confirmed) == 0 {
		break // stable balance for duration of sumPendingTxs
	} // else try again, re-checking all pending tx with transactionConfirmations 
}
```

We'll still need the underflow checks after that since there's no guarantee that the tx confirmations results will reflect balance.

Reply:
Using a loop pattern now.